crates/ruma-state-res/src/events/member.rs

```diff
@@ -0,0 +1,164 @@
+use ruma_events::room::member::MembershipState;
+use serde::Deserialize;
+use serde_json::value::RawValue as RawJsonValue;
+
+use super::Event;
+
+/// A helper type for an [`Event`] of type `m.room.member`.
+///
+/// This is a type that deserializes each field lazily, as requested.
+#[derive(Debug, Clone)]
+pub struct RoomMemberEvent<E: Event>(E);
+
+impl<E: Event> RoomMemberEvent<E> {
+    /// Construct a new `RoomMemberEvent` around the given event.
+    pub fn new(event: E) -> Self {
+        Self(event)
+    }
+
+    /// The membership of the user.
+    pub fn membership(&self) -> Result<MembershipState, String> {
+        RoomMemberEventContent(self.content()).membership()
+    }
+
+    /// If this is a `join` event, the ID of a user on the homeserver that authorized it.
+    pub fn join_authorised_via_users_server(&self) -> Result<Option<OwnedUserId>, String> {
+        RoomMemberEventContent(self.content()).join_authorised_via_users_server()
+    }
+
+    /// If this is an `invite` event, details about the third-party invite that resulted in this
+    /// event.
+    pub(crate) fn third_party_invite(&self) -> Result<Option<ThirdPartyInvite>, String> {
+        RoomMemberEventContent(self.content()).third_party_invite()
+    }
+}
+
+impl<E: Event> Deref for RoomMemberEvent<E> {
+    type Target = E;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// Helper trait for `Option<RoomMemberEvent<E>>`.
+pub(crate) trait RoomMemberEventOptionExt {
+    /// The membership of the user.
+    ///
+    /// Defaults to `leave` if there is no `m.room.member` event.
+    fn membership(&self) -> Result<MembershipState, String>;
+}
+
+impl<E: Event> RoomMemberEventOptionExt for Option<RoomMemberEvent<E>> {
+    fn membership(&self) -> Result<MembershipState, String> {
+        match self {
+            Some(event) => event.membership(),
+            None => Ok(MembershipState::Leave),
+        }
+    }
+}
+
+/// A helper type for the raw JSON content of an event of type `m.room.member`.
+pub(crate) struct RoomMemberEventContent<'a>(&'a RawJsonValue);
+
+impl<'a> RoomMemberEventContent<'a> {
+    /// Construct a new `RoomMemberEventContent` around the given raw JSON content.
+    pub(crate) fn new(content: &'a RawJsonValue) -> Self {
+        Self(content)
+    }
+}
+
+impl RoomMemberEventContent<'_> {
+    /// The membership of the user.
+    pub(crate) fn membership(&self) -> Result<MembershipState, String> {
+        #[derive(Deserialize)]
+        struct RoomMemberContentMembership {
+            membership: Option<MembershipState>,
+        }
+
+        let content: RoomMemberContentMembership =
+            from_raw_json_value(self.0).map_err(|err: serde_json::Error| {
+                format!("invalid `membership` field in `m.room.member` event: {err}")
+            })?;
+        let membership =
+            content.membership.ok_or("missing `membership` field in `m.room.member` event")?;
+        Ok(membership)
```

Comment:
Since we don't even return the `Option`, it seems a bit pointless, since membership events must have a, well,  `membership`
```suggestion
        #[derive(Deserialize)]
        struct RoomMemberContentMembership {
            membership: MembershipState,
        }

        let content: RoomMemberContentMembership =
            from_raw_json_value(self.0).map_err(|err: serde_json::Error| {
                format!("invalid or missing `membership` field in `m.room.member` event: {err}")
            })?;
        Ok(content.membership)
```

Reply:
Applied